### PR TITLE
III-3570 Add Event status

### PR DESCRIPTION
**Pull request by LucWollants**
### Added
 
- Add value object for event status type `EventStatusType`
- Add value object translatable event status reason `EventStatusReason`
- Add domain object for event status containing the type and reason `EventStatus`
 
### Changed
 
- Update `Timestamp` with the new event status domain object
- Update `Calendar` to return the event status in the JSON-LD array

---
Ticket: https://jira.uitdatabank.be/browse/III-3570

### CODE REVIEW

File: src/Timestamp.php [hunk 1]
```diff
@@ -38,7 +38,10 @@ final class Timestamp implements SerializableInterface
 
         $this->startDate = $startDate;
         $this->endDate = $endDate;
-        $this->eventStatus = $eventStatus;
+        $this->eventStatus = new EventStatus(EventStatusType::scheduled(), []);
```
Comment (LucWollants) [inline src/Timestamp.php:41]: Maybe we can use the "Null coalescing operator" here?
```
$this->eventStatus = $eventStatus ?? new EventStatus(EventStatusType::scheduled(), [])
```

Comment (LucWollants) [reply src/Timestamp.php]: Modified to null coalescing operator.

File: src/Calendar.php [hunk 2]
```diff
@@ -227,6 +228,33 @@ public function getTimestamps(): array
         return $this->timestamps;
     }
 
+    public function getEventStatusType(): EventStatusType
+    {
+        $eventStatusTypeCounts = [];
+        $eventStatusTypeCounts[EventStatusType::scheduled()->toNative()] = 0;
+        $eventStatusTypeCounts[EventStatusType::postponed()->toNative()] = 0;
+        $eventStatusTypeCounts[EventStatusType::cancelled()->toNative()] = 0;
```
Comment (jdrieghe) [inline src/Calendar.php:236]: Clever 👏

**Review by jdrieghe — APPROVED** [review]

File: src/Calendar.php (hunk 2 above)
Comment (LucWollants) [reply src/Calendar.php:236]: Although we could miss a new EventStatusType.

Comment (jdrieghe) [reply src/Calendar.php:236]: No, adding a new type starts with adding a test, which will then fail anyway. It's good enough like this.

File: src/Calendar.php [hunk 3]
```diff
@@ -227,6 +228,33 @@ public function getTimestamps(): array
...
+
+        if ($eventStatusTypeCounts[EventStatusType::postponed()->toNative()] > 0) {
+            return EventStatusType::postponed();
+        }
+
+        if ($eventStatusTypeCounts[EventStatusType::cancelled()->toNative()] > 0) {
+            return EventStatusType::cancelled();
+        }
+
+        // This extra return is needed for events with calendar type of permanent or periodic.
```
Comment (bertramakers) [inline src/Calendar.php:254]: Thanks for clarifying with a comment 👍

File: src/Calendar.php (hunk 2 above)
Comment (bertramakers) [reply src/Calendar.php:236]: We would probably miss a new `EventStatusType` with whatever approach to handle this, except by indeed adding a test. But this approach is very readable.

File: src/Event/ValueObjects/EventStatus.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    private $eventStatusType;
+
+    /**
+     * @var EventStatusReason[]
+     */
+    private $eventStatusReasons;
+
+    public function __construct(EventStatusType $eventStatusType, array $eventStatusReasons)
+    {
+        $this->ensureTranslationsAreUnique($eventStatusReasons);
```
Comment (bertramakers) [inline src/Event/ValueObjects/EventStatus.php:23]: Nit: We usually use `guard` I believe, but if you prefer `ensure` that's also okay. It's readable which is the most important. Just wanted to point it out as an FYI.

**Review by bertramakers — APPROVED** [review]
Looks great @LucWollants @jdrieghe, no real comments 👌

**Comment by bertramakers** [comment]
Merged this myself so we can more easily use it in udb3-silex

**Pull request merged**